File: ovstorage-core/examples/python/04_file_and_http.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
"""Route file and HTTP addresses, then compare their capabilities."""

from __future__ import annotations

import argparse
import asyncio
import tempfile
from pathlib import Path

import ovstorage
from ovstorage.file import FileBackend
from ovstorage.plugin import PluginBackend
from ovstorage.redirect_follower import RedirectFollower
from ovstorage.router import Router

from _common import (
    file_connection_request,
    http_connection_request,
    origin_prefix,
    plugin_registry,
)


async def run(url: str, plugin_dir: str | None) -> None:
    with tempfile.TemporaryDirectory(prefix="ovstorage-04-") as directory:
        root = Path(directory)

        # This graph gives one storage handle two address families. The router
        # chooses `files` for file:// and `web` for HTTP(S); the outer redirect
        # follower transparently continues reads when a backend returns a
        # redirect instead of bytes.
        #
        #     redirects (root/wrapper)
        #       └── routes (router)
        #           ├── files (built-in backend)
        #           └── web (HTTP plugin backend)
        storage = await (
            ovstorage.Stack(root="redirects")
            .with_registry(plugin_registry(plugin_dir, "core", "http"))
            .wrapper(RedirectFollower("redirects", "routes"))
            .router(Router("routes", ["files", "web"]))
            .backend(FileBackend("files"))
            .backend(PluginBackend("http", "web"))
            .connection("files", file_connection_request(root, "local files"))
            .connection("web", http_connection_request(origin_prefix(url), "public HTTP"))
            .build()
        )

        # Both URLs enter the same root Layer. Routing is a property of the
        # configured graph, so application code does not select a backend.
        local = (root / "local.txt").as_uri()
        await storage.write(local, b"local data\n")
        print("file read:", (await storage.read_bytes(local))[0].decode().rstrip())
        print("http read:", len((await storage.read_bytes(url, max_bytes=1024 * 1024))[0]), "bytes")

        # Backends advertise different capabilities. A filesystem has
        # directory listings; this HTTP backend models individual web objects
        # and reports Unsupported for list rather than emulating a directory.
        page = await storage.list(root.as_uri() + "/")
        print("file list:", [item.address for item in page.items])
        try:
            await storage.list(origin_prefix(url))
        except ovstorage.UnsupportedError as error:
            print(f"http list: {error.code} (HTTP addresses are objects, not directories)")
        else:
            raise SystemExit("expected the HTTP backend to reject list with Unsupported")


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "url",
        nargs="?",
        default="https://raw.githubusercontent.com/NVIDIA-Omniverse/ovstorage/main/README.md",
    )
    parser.add_argument("--plugin-dir")
    args = parser.parse_args()
    asyncio.run(run(args.url, args.plugin_dir))


if __name__ == "__main__":
    main()

File: ovstorage-c-source/examples/tutorial/04_file_and_http.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

#include "common.hpp"

int main(int argc, char** argv)
{
    if (argc < 2 || argc > 3) {
        std::cerr << "usage: " << argv[0] << " PLUGIN_DIR [HTTP_URL]\n";
        return 2;
    }
    const std::string url =
        argc == 3 ? argv[2]
                  : "https://raw.githubusercontent.com/"
                    "NVIDIA-Omniverse/ovstorage/main/README.md";
    tutorial::Context context;
    // Core supplies Router and RedirectFollower; HTTP supplies the web backend.
    // Context retains both plugin objects for the lifetime of the built graph.
    if (!context.load(argv[1], "core") || !context.load(argv[1], "http")) {
        return 1;
    }

    tutorial::TempDirectory work("ovstorage-cpp-04");
    if (!work) return 1;
    const auto& directory = work.path();
    ovstorage::Stack stack;
    if (!tutorial::add_file(stack, context, "files", directory) ||
        !tutorial::add_http(stack, context, "web", url)) return 1;
    // One router gives the handle two address families. It selects `files` for
    // file:// and `web` for HTTP(S), based on each connection's declared root.
    if (!tutorial::ok(
            "Stack::add_layer(router)",
            stack.add_layer(context.registry, "routes", "router"))) return 1;
    // The outer wrapper follows redirect read results before returning bytes:
    //
    //     redirects (root/wrapper)
    //       `-- routes (router)
    //           |-- files (built-in backend)
    //           `-- web (HTTP plugin backend)
    if (!tutorial::ok(
            "Stack::set_children",
            stack.set_children("routes", {"files", "web"}))) return 1;
    if (!tutorial::ok(
            "Stack::add_layer(redirect follower)",
            stack.add_layer(context.registry, "redirects", "redirect_follower")) ||
        !tutorial::ok(
            "Stack::set_inner",
            stack.set_inner("redirects", "routes")) ||
        !tutorial::ok(
            "Stack::set_root",
            stack.set_root("redirects"))) return 1;

    auto built = ovstorage::sync_wait(stack.build());
    if (!tutorial::ok("Stack::build", built)) return 1;
    auto storage = std::move(built).value();

    // Both URLs enter the same Stack root; application code never names the
    // backend that ultimately serves an operation.
    auto remote = ovstorage::sync_wait(storage.read_bytes(url));
    if (!tutorial::ok("HTTP read", remote)) return 1;
    std::cout << "HTTP read: " << remote.value().first.span().size()
              << " bytes\n";

    // Capabilities belong to the selected backend. Filesystems can list a
    // directory, while this HTTP backend models individual objects and reports
    // Unsupported instead of inventing directory semantics.
    auto file_list =
        ovstorage::sync_wait(storage.list(tutorial::file_root(directory)));
    if (!tutorial::ok("file list", file_list)) return 1;
    auto http_list = ovstorage::sync_wait(storage.list(tutorial::origin(url)));
    if (http_list || http_list.error().code() != OvStorage_Status_Unsupported) {
        std::cerr << "HTTP list should return Unsupported\n";
        return 1;
    }
    std::cout << "HTTP list: Unsupported\n";
    return 0;
}
